stock_inventory_merge/models/stock_inventory.py

```diff
@@ -109,14 +109,18 @@ def action_merge_duplicated_line(self):
                                 uom_obj.browse(default_uom_id),
                             )
 
-                # Update the first line with the sumed quantity
-                keeped_line = line_obj.browse(keeped_line_id)
-                keeped_line.write({"product_qty": sum_quantity})
-
                 # Delete all the other lines
                 line_ids.remove(keeped_line_id)
                 line_obj.browse(line_ids).unlink()
 
+                # Update the first line with the sumed quantity
+                # Note: This has to be done *after* deleting the other lines,
+                # because the `write()` calls the `_check_no_duplicate_line()`
```

Comment:
it should not fails, due to that https://github.com/grap/grap-odoo-incubator/blob/12.0/stock_inventory_merge/models/stock_inventory_line.py#L25

could you explain in which case it is failing. i don't face any problem (and I have a lot of duplicates !)

Reply:
The issue here is that if the update `keeped_line.write(...)` comes before the removal `line_obj.browse(line_ids).unlink()`, then the duplicates are still there during the update, and the Odoo core method `write()` will do a check for duplicates and thus -- under certain conditions -- raise the exception:
https://github.com/OCA/OCB/blob/a082f6d9ae716561a40df61a5f77e1c089a97945/addons/stock/models/stock_inventory.py#L405-L409
Note that there the condition used in the method `_check_no_duplicate_line()` for actual checking does not simply compare the `product_id`, but a bit mode, including a check for `inventory_id.state` being `'confirm'`.
So, as long as the inventory is not yet confirmed, you should be fine. I'm not sure if the inventory being already confirmed is "too artificial" a condition, and usually won't occur in practice.

Reply:
Hi,
sorry, I was a bit confused here myself, especially about when this failing check happens. Let's see...

First of all, state `'confirm'` is the "In Progess" state, which is certainly not "too artificial", but rather the expected state for a started, but not yet validated (a.k.a. done or finalized) inventory. (The state `'done'` = "Validated" would be "too artificial" for the merging of inventories, in the sense that such done inventories should never be merged anymore. But this case is not what I had in mind, anyway.)

Further, module "stock_inventory_merge" [provides a flag `do_not_check_duplicates=True` in the context](https://github.com/grap/grap-odoo-incubator/blob/ddccb93400161d152ab3603cfc6cf2f95c4ef327/stock_inventory_merge/models/stock_inventory_line.py#L14-L20) when creating or writing "stock.inventory.line" objects.
Now, the [check `_check_no_duplicate_line()`](https://github.com/odoo/odoo/blob/ffca224ebe3731ffdff2f18fd0b340f904ff4187/addons/stock/models/stock_inventory.py#L409-L422), which is done in core module "stock" when creating or writing a "stock.inventory.line" record, does a `search()` for duplicate lines. However, module "stock_inventory_merge" [overwrites the search domain when this context flag `do_not_check_duplicates` is `True`](https://github.com/grap/grap-odoo-incubator/blob/ddccb93400161d152ab3603cfc6cf2f95c4ef327/stock_inventory_merge/models/stock_inventory_line.py#L25-L26) -- to the simple domain `[('product_id', '=', False)]`, which usually will yield no matching records.
This essentially means that when the module "stock_inventory_merge" is installed, the check for duplicate lines is "suppressed" when creating or writing "stock.inventory.line" records.
So, the actual merging of the inventories should be fine.

However, the method `action_merge_duplicated_line()` we are talking about here is executed when merging duplicate lines of the merged inventory via button "Merge Duplicates".
And here when clicking on that button the [exception "You cannot have two inventory adjustments..."](https://github.com/odoo/odoo/blob/ffca224ebe3731ffdff2f18fd0b340f904ff4187/addons/stock/models/stock_inventory.py#L420-L422) will be raised if there are multiple lines with the same product, location etc.
And _of course_ this is just the situation when the button should be clicked.

So, if the [accumulated quantity is written](https://github.com/grap/grap-odoo-incubator/blob/ddccb93400161d152ab3603cfc6cf2f95c4ef327/stock_inventory_merge/models/stock_inventory.py#L114) to the line to be kept (`keeped_line`) before [removing the duplicate lines](https://github.com/grap/grap-odoo-incubator/blob/ddccb93400161d152ab3603cfc6cf2f95c4ef327/stock_inventory_merge/models/stock_inventory.py#L116-L118), this `write()` will execute the check `_check_no_duplicate_line()`, this time without the context flag `do_not_check_duplicates`. And thus the exception is raised.
In other words, the method `action_merge_duplicated_line()` for button "Merge Duplicates" should never work when it is supposed to do (i.e., when there actually are duplicate lines in the inventory).

Summarized, I think this commit https://github.com/grap/grap-odoo-incubator/commit/93ebd394c00ac8a9c8cf1426a3a573baf89c35d7 is actually necessary.
Alternatively, one could provide the flag `do_not_check_duplicates=True` in the context when [writing the accumulated quantity to the kept line](https://github.com/grap/grap-odoo-incubator/blob/ddccb93400161d152ab3603cfc6cf2f95c4ef327/stock_inventory_merge/models/stock_inventory.py#L114), i.e.,
```python
keeped_line.with_context(do_not_check_duplicates=True).write({"product_qty": sum_quantity})
```
Do you think this would be a cleaner solution?